File: pkg/strategy/decide.go
package strategy

import (
	"fmt"
	"math/big"

	"github.com/ethereum/go-ethereum/common"

	"github.com/everstrat-xyz/keepers/pkg/solmath"
)

// Scan bounds from CREStrategyExecutor. W2 mirrors them deliberately — see
// Decide.
const (
	// MaxBatchScan is CREStrategyExecutor.MAX_BATCH_SCAN.
	MaxBatchScan = 25
	// MaxUsersCostScan is CREStrategyExecutor.MAX_USERS_COST_SCAN.
	MaxUsersCostScan = 50
)

// Strategy is one registered strategy as read off-chain.
type Strategy struct {
	Address common.Address
	Paused  bool
	Healthy bool
	// MaxDeposit and MaxWithdrawal are capacity, not balances.
	MaxDeposit    *big.Int
	MaxWithdrawal *big.Int
	// InDepositCooldown comes from StrategyManager, not the strategy.
	InDepositCooldown bool
	// DepositWeight comes from StrategyManager. A registered-but-unfunded
	// strategy has weight 0 and can never take a deposit, so
	// `_depositCapacityAvailable` requires it to be non-zero (contracts PR
	// #43, R4-M-04) — without the gate W2 would recommend a DepositExcess the
	// StrategyManager always no-ops.
	DepositWeight uint8
	// PendingPerformanceFeeETH is this strategy's accrued fee.
	PendingPerformanceFeeETH *big.Int
}

// State is the snapshot W2 decides from.
type State struct {
	// Now is the observed block's timestamp — the same clock lastSyncAt was
	// recorded with. See docs/envelope.md.
	Now uint64

	// Paused is true when the receiver or Controller/StrategyManager is
	// paused; both `strategyUpkeepStatus` and every action gate on it.
	Paused bool

	ControllerBalance *big.Int

	// NeedsETH is the pending redemption cost, computed with the *same*
	// bounded scan the contract uses. See PendingRedemptionNeedsETH.
	NeedsETH *big.Int

	// AMMFreeBalance is the exit-side float.
	AMMFreeBalance *big.Int

	Strategies []Strategy

	// PerformanceFeeBps is zero when fees are disabled, which short-circuits
	// the harvest branch.
	PerformanceFeeBps *big.Int

	// Thresholds, mirroring the receiver's storage.
	ControllerReserveETH     *big.Int
	MinDepositETH            *big.Int
	MinWithdrawETH           *big.Int
	MinHarvestETH            *big.Int
	ExitLiquidityTargetETH   *big.Int
	MinExitLiquidityTopUpETH *big.Int
	SyncInterval             uint64
	LastSyncAt               uint64

	// ScanTruncated records that the read budget could not cover the full
	// bounded scan, so NeedsETH may understate the true figure.
	ScanTruncated bool
}

// Decision is the action W2 proposes.
//
// Amount is diagnostic only — it is logged, never reported. The receiver
// recomputes every quantity at execution time, which is why the report carries
// no amount at all.
type Decision struct {
	Action Action
	Amount *big.Int
	Reason string
}

// Decide selects the action, mirroring `CREStrategyExecutor.strategyUpkeepStatus`
// branch for branch and in the same order.
//
// # Why this deliberately does not compute a more exact shortfall
//
// Issue #4 framed W2's advantage as an "exact shortfall vs truncated on-chain
// scan". That would be actively harmful here, and the reason is worth stating:
// `_processReport` re-derives every quantity with the *same* bounded helpers
// the view uses (MAX_BATCH_SCAN batches, MAX_USERS_COST_SCAN users). A workflow
// that computed a truer shortfall would propose `WithdrawShortfall` in states
// where the receiver's own recomputation sees no shortfall — and revert with
// `KeeperExecutorNoUpkeepNeeded` every time.
//
// W1 is the opposite case: `_processReport` validates a `ProcessRequests` claim
// per batch, with no window, so scanning deeper than the view genuinely wins.
// The asymmetry is in the contracts, not in the workflows.
//
// So W2's value is not precision. It is running the decision against live state
// off-chain, cross-checking the view, and being the thing that can actually
// deliver a report.
func Decide(s State) (Decision, error) {
	zero := new(big.Int)

	if s.Paused {
		return Decision{Action: ActionNone, Amount: zero, Reason: "receiver or a protocol contract is paused"}, nil
	}

	// 1. Rebalance — any live strategy that is unhealthy.
	for _, st := range s.Strategies {
		if !st.Paused && !st.Healthy {
			return Decision{
				Action: ActionRebalance,
				Amount: zero,
				Reason: fmt.Sprintf("strategy %s is unhealthy and not paused", st.Address),
			}, nil
		}
	}

	needs := orZero(s.NeedsETH)
	balance := orZero(s.ControllerBalance)

	// 2. WithdrawShortfall — pending redemptions exceed the Controller's
	// balance by at least minWithdrawETH, and some strategy can actually pay.
	if needs.Cmp(balance) > 0 {
		shortfall := new(big.Int).Sub(needs, balance)
		if shortfall.Cmp(orZero(s.MinWithdrawETH)) >= 0 && s.totalMaxWithdrawal().Sign() > 0 {
			return Decision{
				Action: ActionWithdrawShortfall,
				Amount: shortfall,
				Reason: fmt.Sprintf("pending redemptions need %s wei against a controller balance of %s wei",
					needs, balance),
			}, nil
		}
	}

	// 3. ProvideExitLiquidity — top the AMM float back toward its target.
	topUp := s.exitLiquidityTopUp(balance, needs)
	if topUp.Cmp(orZero(s.MinExitLiquidityTopUpETH)) >= 0 && topUp.Sign() > 0 {
		return Decision{
			Action: ActionProvideExitLiquidity,
			Amount: topUp,
			Reason: fmt.Sprintf("AMM float %s wei is below the %s wei target; topping up by %s wei",
				orZero(s.AMMFreeBalance), orZero(s.ExitLiquidityTargetETH), topUp),
		}, nil
	}

	// 4. DepositExcess — idle ETH above the reserve, with somewhere to put it.
	excess := s.idleExcess(balance, needs)
	if excess.Cmp(orZero(s.MinDepositETH)) >= 0 && s.depositCapacityAvailable() {
		return Decision{
			Action: ActionDepositExcess,
			Amount: excess,
			Reason: fmt.Sprintf("controller holds %s wei idle above the reserve and pending redemptions", excess),
		}, nil
	}

	// 5. HarvestPerformanceFees.
	fee := s.pendingPerformanceFeeETH()
	if fee.Cmp(orZero(s.MinHarvestETH)) >= 0 && fee.Sign() > 0 {
		return Decision{
			Action: ActionHarvestPerformanceFees,
			Amount: fee,
			Reason: fmt.Sprintf("accrued performance fees are %s wei", fee),
		}, nil
	}

	// 6. Sync — periodic accounting refresh.
	if s.SyncInterval != 0 && len(s.Strategies) > 0 && s.Now >= s.LastSyncAt {
		if elapsed := s.Now - s.LastSyncAt; elapsed >= s.SyncInterval {
			return Decision{
				Action: ActionSync,
				Amount: zero,
				Reason: fmt.Sprintf("%ds since the last sync (interval %ds)", elapsed, s.SyncInterval),
			}, nil
		}
	}

	return Decision{Action: ActionNone, Amount: zero, Reason: "no strategy action is due"}, nil
}

// totalMaxWithdrawal mirrors _totalMaxWithdrawal.
func (s State) totalMaxWithdrawal() *big.Int {
	total := new(big.Int)
	for _, st := range s.Strategies {
		total.Add(total, orZero(st.MaxWithdrawal))
	}
	return total
}

// depositCapacityAvailable mirrors _depositCapacityAvailable.
//
// depositWeight > 0 is the R4-M-04 gate: all-zero weights are
// registered-but-unfunded — the StrategyManager skips them and refunds the
// Controller, so proposing DepositExcess against only those is work the
// receiver reverts on (the view says none is available).
func (s State) depositCapacityAvailable() bool {
	for _, st := range s.Strategies {
		if !st.InDepositCooldown && st.Healthy &&
			orZero(st.MaxDeposit).Sign() > 0 && st.DepositWeight > 0 {
			return true
		}
	}
	return false
}

// pendingPerformanceFeeETH mirrors _pendingPerformanceFeeETH, including the
// short-circuit when fees are disabled.
func (s State) pendingPerformanceFeeETH() *big.Int {
	total := new(big.Int)
	if orZero(s.PerformanceFeeBps).Sign() == 0 {
		return total
	}
	for _, st := range s.Strategies {
		total.Add(total, orZero(st.PendingPerformanceFeeETH))
	}
	return total
}

// idleExcess mirrors _idleExcess: balance above the reserve plus what pending
// redemptions have already claimed.
func (s State) idleExcess(balance, needs *big.Int) *big.Int {
	reserved := new(big.Int).Add(orZero(s.ControllerReserveETH), needs)
	if balance.Cmp(reserved) <= 0 {
		return new(big.Int)
	}
	return new(big.Int).Sub(balance, reserved)
}

// exitLiquidityTopUp mirrors _exitLiquidityTopUp: the smaller of the float
// shortfall and the idle excess.
func (s State) exitLiquidityTopUp(balance, needs *big.Int) *big.Int {
	target := orZero(s.ExitLiquidityTargetETH)
	if target.Sign() == 0 {
		return new(big.Int)
	}
	float := orZero(s.AMMFreeBalance)
	if float.Cmp(target) >= 0 {
		return new(big.Int)
	}

	shortfall := new(big.Int).Sub(target, float)
	excess := s.idleExcess(balance, needs)
	if shortfall.Cmp(excess) < 0 {
		return shortfall
	}
	return excess
}

// QueueBatch is the subset of a batch needed to price pending redemptions.
//
// TotalTokensToBurn is deliberately absent: the current batch's unpriced EVE is
// not a liability until `priceBatch` (contracts PR #43, M-11), so W2 never
// needs to read it.
type QueueBatch struct {
	ID               uint64
	CanBeProcessed   bool
	FinalEvePrice    *big.Int
	PricedAt         uint64
	UnprocessedCount uint64
	Requests         []QueueRequest
}

// QueueRequest is one unprocessed redemption request.
type QueueRequest struct {
	EvePriceAtRequestTime *big.Int
	TokensToBurn          *big.Int
	PriceTolerance        *big.Int
}

// PendingRedemptionNeedsETH mirrors `_pendingRedemptionNeedsETH`.
//
// cursor is the queue keeper's `nextLiveBatchIdToProcess()`. The walk is capped
// at MaxBatchScan batches and MaxUsersCostScan users per batch — matching the
// contract exactly, because the receiver re-derives this figure the same way
// when it validates the report. Widening the caps here would produce reports
// the receiver rejects.
//
// The current batch is deliberately absent: until `priceBatch` it is still
// cancellable equity (`liveRedemptionOffsets` is zero), so the contract counts
// it as no liability at all. Sizing it at the live base price here would make
// the workflow propose WithdrawShortfall in states where the receiver's own
// recomputation sees none — and every report would revert with
// KeeperExecutorNoUpkeepNeeded.
func PendingRedemptionNeedsETH(
	batches map[uint64]QueueBatch,
	cursor, currentBatchID uint64,
	maxBatchProcessingTime, now uint64,
) (*big.Int, error) {
	needs := new(big.Int)

	limit := cursor + MaxBatchScan
	for id := cursor; id < currentBatchID && id < limit; id++ {
		b, ok := batches[id]
		if !ok {
			continue
		}
		cost, err := batchSettlementCost(b, maxBatchProcessingTime, now)
		if err != nil {
			return nil, err
		}
		needs.Add(needs, cost)
	}
	return needs, nil
}

// batchSettlementCost mirrors _batchSettlementCost.
//
// Note the difference from W1's affordability walk: this sums every request's
// cost with no balance budget and no early break, because it is asking "what do
// pending redemptions cost?" rather than "what can we afford right now?".
func batchSettlementCost(b QueueBatch, maxBatchProcessingTime, now uint64) (*big.Int, error) {
	cost := new(big.Int)
	if !b.CanBeProcessed {
		return cost, nil
	}
	// A batch past the escape hatch is users' to close, not the keeper's.
	if b.PricedAt > 0 && now > b.PricedAt+maxBatchProcessingTime {
		return cost, nil
	}
	if b.UnprocessedCount == 0 {
		return cost, nil
	}

	limit := b.UnprocessedCount
	if limit > MaxUsersCostScan {
		limit = MaxUsersCostScan
	}
	if uint64(len(b.Requests)) < limit {
		limit = uint64(len(b.Requests))
	}

	for i := uint64(0); i < limit; i++ {
		r := b.Requests[i]
		slipped, err := solmath.IsRelativelyLessThan(
			orZero(b.FinalEvePrice), orZero(r.EvePriceAtRequestTime), orZero(r.PriceTolerance))
		if err != nil {
			return nil, fmt.Errorf("batch %d request %d: %w", b.ID, i, err)
		}
		if slipped {
			continue
		}
		cost.Add(cost, solmath.ConvertAssets(orZero(r.TokensToBurn), orZero(b.FinalEvePrice)))
	}
	return cost, nil
}

func orZero(v *big.Int) *big.Int {
	if v == nil {
		return new(big.Int)
	}
	return v
}

File: mimic-functions/queue-keeper/src/decide.ts
/**
 * W1 queue-keeper decision engine — ported from the CRE-era Go implementation
 * (`pkg/queue/decide.go`) with its semantics intact.
 *
 * The engine decides what `QueueKeeperExecutor.perform` should be called with,
 * and the params surface here stays closed: the only expressible values are a
 * batch id and an index range. There is no path to put an ETH amount, NAV, or
 * price into a W1 payload, and `decodeParams` rejects any blob longer than the
 * action's exact encoding — which is what an appended amount word would look
 * like on the wire. The executor re-derives everything from live state anyway;
 * the discipline exists so a future edit cannot smuggle an authoritative value
 * into a payload and start trusting it.
 */

import { BigInt } from '@mimicprotocol/lib-ts'

import { convertAssets, isRelativelyLessThan } from './solmath'

/**
 * `QueueKeeperExecutor.MAX_BATCH_SCAN` — the gas bound on the on-chain
 * `queueUpkeepStatus` view and on `_advanceBatchCursor`.
 *
 * W1 deliberately scans past it (see decide). The executor's `_processReport`
 * does *not* apply the window when validating a `ProcessRequests` claim, so a
 * batch found beyond it is still executable — that is the whole point of
 * scanning off-chain. `AdvanceCursor` is the exception, and decide caps that
 * claim accordingly.
 */
export const MAX_BATCH_SCAN: BigInt = BigInt.fromI32(25)

/** QueueAction values from `IQueueKeeperExecutor`. */
export const ActionNone: u8 = 0
export const ActionPriceBatch: u8 = 1
export const ActionProcessRequests: u8 = 2
export const ActionAdvanceCursor: u8 = 3

export function actionString(a: u8): string {
  if (a == ActionNone) return 'None'
  if (a == ActionPriceBatch) return 'PriceBatch'
  if (a == ActionProcessRequests) return 'ProcessRequests'
  if (a == ActionAdvanceCursor) return 'AdvanceCursor'
  return 'Action(' + a.toString() + ')'
}

/** One queued redemption, as read from `ExitQueue.requestInfo`. */
export class Request {
  user: string
  processed: bool
  closedDueToSlippage: bool
  evePriceAtRequestTime: BigInt
  tokensToBurn: BigInt
  priceTolerance: BigInt

  constructor(
    user: string,
    processed: bool,
    closedDueToSlippage: bool,
    evePriceAtRequestTime: BigInt,
    tokensToBurn: BigInt,
    priceTolerance: BigInt
  ) {
    this.user = user
    this.processed = processed
    this.closedDueToSlippage = closedDueToSlippage
    this.evePriceAtRequestTime = evePriceAtRequestTime
    this.tokensToBurn = tokensToBurn
    this.priceTolerance = priceTolerance
  }
}

/** One ExitQueue batch: `batchInfo` plus the unprocessed user list. */
export class Batch {
  id: BigInt
  canBeProcessed: bool
  finalEvePrice: BigInt
  totalTokensToBurn: BigInt
  createdAt: BigInt
  pricedAt: BigInt
  unprocessedCount: BigInt

  /**
   * Unprocessed requests in `unprocessedUsers` order, starting at index 0. The
   * executor only accepts a prefix of this list, so order is load-bearing, not
   * incidental. May be shorter than unprocessedCount when the read was capped
   * at maxUsersPerUpkeep — which is all the affordability model needs, since
   * the executor caps there too.
   */
  requests: Request[]

  constructor(
    id: BigInt,
    canBeProcessed: bool,
    finalEvePrice: BigInt,
    totalTokensToBurn: BigInt,
    createdAt: BigInt,
    pricedAt: BigInt,
    unprocessedCount: BigInt,
    requests: Request[]
  ) {
    this.id = id
    this.canBeProcessed = canBeProcessed
    this.finalEvePrice = finalEvePrice
    this.totalTokensToBurn = totalTokensToBurn
    this.createdAt = createdAt
    this.pricedAt = pricedAt
    this.unprocessedCount = unprocessedCount
    this.requests = requests
  }
}

/** The full off-chain snapshot W1 decides from — one tick, no memory. */
export class State {
  now: BigInt
  /** True when the executor or any of Controller / ExitQueue / AMM is paused. */
  paused: bool
  currentBatchId: BigInt
  /** The executor's stored cursor (`nextBatchIdToProcess`). */
  nextBatchIdToProcess: BigInt
  /** The ETH budget redemptions are paid from — the Controller's balance. */
  controllerBalance: BigInt
  maxBatchProcessingTime: BigInt
  minBatchAge: BigInt
  maxUsersPerUpkeep: BigInt
  /** Full scan, keyed by batch id as a decimal string. */
  batches: Map<string, Batch>
  /**
   * Last batch id the read layer actually fetched when it capped the scan.
   * Zero means the scan reached the current batch. A truncated scan can only
   * cause W1 to propose *less* work than exists, never wrong work — but it
   * makes "found nothing" ambiguous, so decide refuses PriceBatch until a
   * later tick can finish the walk.
   */
  scanTruncatedAt: BigInt

  constructor(
    now: BigInt,
    paused: bool,
    currentBatchId: BigInt,
    nextBatchIdToProcess: BigInt,
    controllerBalance: BigInt,
    maxBatchProcessingTime: BigInt,
    minBatchAge: BigInt,
    maxUsersPerUpkeep: BigInt,
    batches: Map<string, Batch>,
    scanTruncatedAt: BigInt
  ) {
    this.now = now
    this.paused = paused
    this.currentBatchId = currentBatchId
    this.nextBatchIdToProcess = nextBatchIdToProcess
    this.controllerBalance = controllerBalance
    this.maxBatchProcessingTime = maxBatchProcessingTime
    this.minBatchAge = minBatchAge
    this.maxUsersPerUpkeep = maxUsersPerUpkeep
    this.batches = batches
    this.scanTruncatedAt = scanTruncatedAt
  }
}

export class Decision {
  action: u8
  batchId: BigInt
  /** Exclusive end of the claimed affordable prefix (ProcessRequests only). */
  endIndex: BigInt
  reason: string
  /**
   * The chosen batch sits past MAX_BATCH_SCAN from the executor's stored
   * cursor — i.e. the on-chain view structurally could not have found it.
   * Divergence classification uses this to separate a real improvement from
   * a bug.
   */
  scannedBeyondWindow: bool

  constructor(action: u8, batchId: BigInt, endIndex: BigInt, reason: string, scannedBeyondWindow: bool) {
    this.action = action
    this.batchId = batchId
    this.endIndex = endIndex
    this.reason = reason
    this.scannedBeyondWindow = scannedBeyondWindow
  }
}

export function getBatch(s: State, id: BigInt): Batch | null {
  // AssemblyScript's Map.get aborts the whole module on a missing key, and a
  // truncated scan leaves ids past scanEnd absent — so probe membership first.
  const key = id.toString()
  if (!s.batches.has(key)) return null
  return s.batches.get(key)
}

export function scanTruncated(s: State): bool {
  return s.scanTruncatedAt != BigInt.zero() && s.scanTruncatedAt < s.currentBatchId
}

/**
 * Mirrors `QueueKeeperExecutor._isBatchSkippable`.
 *
 * The unpriced guard comes first, so an unpriced batch — including the current
 * one, even when still empty — is never skippable: it can still receive
 * requests and must be priced first. A priced batch is skippable when fully
 * processed, or past MAX_BATCH_PROCESSING_TIME (the escape hatch, after which
 * users close their own requests and the keeper must not touch the batch).
 *
 * An unreadable batch is *not* skippable: skipping past a batch we could not
 * read would advance the cursor over live work.
 */
export function isBatchSkippable(s: State, id: BigInt): bool {
  const b = getBatch(s, id)
  if (b == null) return false
  if (!b.canBeProcessed) return false
  if (b.unprocessedCount == BigInt.zero()) return true
  return s.now > b.pricedAt.plus(s.maxBatchProcessingTime)
}

/**
 * Mirrors `QueueKeeperExecutor._affordableRequests`: how many requests, taken
 * as a prefix from index 0, the Controller's balance covers.
 *
 * The contract walks the prefix accumulating cost and stops at the first
 * request that would overrun the balance — it does not skip an expensive
 * request to fit cheaper ones behind it. Reproducing the break (rather than
 * "fit as many as possible") is what keeps the claim acceptable to the
 * executor.
 *
 * A batch past its processing window returns zero: the executor's own walk
 * returns zero there and the batch is the users' to close, not the keeper's.
 */
export function affordableRequests(s: State, id: BigInt): BigInt {
  const b = getBatch(s, id)
  if (b == null) return BigInt.zero()
  if (!b.canBeProcessed) return BigInt.zero()
  if (b.pricedAt > BigInt.zero() && s.now > b.pricedAt.plus(s.maxBatchProcessingTime)) return BigInt.zero()
  if (b.unprocessedCount == BigInt.zero()) return BigInt.zero()

  let limit = b.unprocessedCount
  if (limit > s.maxUsersPerUpkeep) limit = s.maxUsersPerUpkeep
  if (BigInt.fromI32(b.requests.length) < limit) limit = BigInt.fromI32(b.requests.length)

  let cumulative = BigInt.zero()
  let count = BigInt.zero()
  for (let i = BigInt.zero(); i < limit; i = i.plus(BigInt.fromI32(1))) {
    const cost = requestCost(b.finalEvePrice, b.requests[i.toI32()])
    const next = cumulative.plus(cost)
    if (next > s.controllerBalance) break
    cumulative = next
    count = count.plus(BigInt.fromI32(1))
  }
  return count
}

/**
 * Mirrors the ETH a single request costs the Controller, per
 * `Controller._processRequest`. A request whose batch price fell more than the
 * user's tolerance below their queued price is closed at zero cost — it
 * consumes a slot but no ETH.
 */
export function requestCost(finalEvePrice: BigInt, r: Request): BigInt {
  if (isRelativelyLessThan(finalEvePrice, r.evePriceAtRequestTime, r.priceTolerance)) {
    return BigInt.zero()
  }
  return convertAssets(r.tokensToBurn, finalEvePrice)
}

/**
 * The first batch id `queueUpkeepStatus` cannot reach. The view walks its own
 * bounded cursor first, then scans at most MAX_BATCH_SCAN batches from wherever
 * that lands — so its reach is two windows deep from the stored cursor, not
 * one. Both decide and classify go through this helper so "did the view have a
 * chance to see this?" has exactly one answer.
 */
export function onChainScanWindowEnd(s: State): BigInt {
  return peekAdvancedCursor(s, MAX_BATCH_SCAN).plus(MAX_BATCH_SCAN)
}

/**
 * Mirrors `QueueKeeperExecutor._peekAdvancedCursor`. scanLimit bounds how far
 * the cursor may walk: pass MAX_BATCH_SCAN to reproduce what the executor will
 * do; pass zero for an unbounded off-chain walk.
 */
export function peekAdvancedCursor(s: State, scanLimit: BigInt): BigInt {
  let cursor = s.nextBatchIdToProcess
  let stop = s.currentBatchId
  if (scanLimit > BigInt.zero() && cursor.plus(scanLimit) < stop) {
    stop = cursor.plus(scanLimit)
  }
  while (cursor < stop && isBatchSkippable(s, cursor)) {
    cursor = cursor.plus(BigInt.fromI32(1))
  }
  return cursor
}

/**
 * Decide selects the action W1 proposes, following the same priority order as
 * `QueueKeeperExecutor.queueUpkeepStatus` — process work first, then price the
 * current batch, then advance the cursor.
 *
 * The deliberate difference from the on-chain view is the scan width: this
 * walks every batch from the cursor to the current one, where the view stops
 * after MAX_BATCH_SCAN. That is the improvement the off-chain keeper exists to
 * provide, and it is safe for `ProcessRequests` because `_processReport`
 * re-derives affordability for the named batch without applying the window.
 *
 * `AdvanceCursor` gets the opposite treatment: the executor advances its
 * cursor with the *bounded* walk, so a claim past `cursor + MAX_BATCH_SCAN` is
 * unreachable in one execution and reverts. Decide therefore claims only the
 * bounded cursor.
 */
export function decide(s: State): Decision {
  if (s.paused) {
    return none('executor or a protocol contract is paused')
  }

  const fullCursor = peekAdvancedCursor(s, BigInt.zero())
  const windowEnd = onChainScanWindowEnd(s)

  // Process the oldest batch with an affordable prefix. Full scan.
  for (let id = fullCursor; id < s.currentBatchId; id = id.plus(BigInt.fromI32(1))) {
    if (isBatchSkippable(s, id)) continue
    const affordable = affordableRequests(s, id)
    if (affordable == BigInt.zero()) continue
    return new Decision(
      ActionProcessRequests,
      id,
      affordable,
      'batch ' +
        id.toString() +
        ' has ' +
        affordable.toString() +
        ' affordable requests within a controller balance of ' +
        s.controllerBalance.toString() +
        ' wei',
      id >= windowEnd
    )
  }

  // PriceBatch is accepted even when ProcessRequests was also due, and would
  // grow the live-priced set instead of settling. If the process walk did not
  // finish, we cannot know that nothing was affordable — skip pricing this
  // tick. AdvanceCursor is still safe: it only walks skippable batches we
  // already have headers for.
  if (!scanTruncated(s)) {
    const b = getBatch(s, s.currentBatchId)
    if (b != null && b.unprocessedCount > BigInt.zero() && s.now >= b.createdAt) {
      const age = s.now.minus(b.createdAt)
      if (age >= s.minBatchAge) {
        return new Decision(
          ActionPriceBatch,
          s.currentBatchId,
          BigInt.zero(),
          'current batch ' +
            s.currentBatchId.toString() +
            ' has ' +
            b.unprocessedCount.toString() +
            ' unprocessed requests and is ' +
            age.toString() +
            's old (minBatchAge ' +
            s.minBatchAge.toString() +
            's)',
          false
        )
      }
    }
  }

  // Advance the cursor past dead batches — capped at what the executor can
  // actually reach in one execution.
  const boundedCursor = peekAdvancedCursor(s, MAX_BATCH_SCAN)
  if (boundedCursor > s.nextBatchIdToProcess) {
    return new Decision(
      ActionAdvanceCursor,
      boundedCursor,
      BigInt.zero(),
      'cursor can advance from ' +
        s.nextBatchIdToProcess.toString() +
        ' to ' +
        boundedCursor.toString() +
        ' past fully-processed or expired batches',
      false
    )
  }

  return none('no batch needs pricing, processing, or a cursor advance')
}

function none(reason: string): Decision {
  return new Decision(ActionNone, BigInt.zero(), BigInt.zero(), reason, false)
}
